File: server/src/main/kotlin/com/duluin/ftth/inventory/domain/model/WorkOrderMaterial.kt
package com.duluin.ftth.inventory.domain.model

import com.duluin.ftth.common.domain.UuidV7
import com.duluin.ftth.common.domain.error.ConflictException
import com.duluin.ftth.common.domain.error.ValidationException
import java.time.Instant
import java.util.UUID

/**
 * Nasib satu unit berserial yang sudah keluar gudang untuk sebuah work order.
 *
 * Ketiganya WAJIB dideklarasikan — tidak ada pilihan "tidak tahu". Unit yang sudah dipegang
 * teknisi tapi tidak pernah dilaporkan nasibnya adalah kebocoran aset yang paling sulit
 * dilihat: saldo teknisi terus menggelembung, dan setiap opname akan menyalahkan orang yang
 * berbeda-beda.
 */
enum class MaterialOutcome {
    /** Terpasang di pelanggan. Inilah satu-satunya nasib yang MEMOTONG saldo saat WO disetujui. */
    INSTALLED,

    /** Dibawa pulang utuh. Masih harus benar-benar diterima gudang lewat retur biasa. */
    RETURNED,

    /** Hilang atau rusak di lapangan. Harus lewat penyesuaian stok yang butuh persetujuan. */
    LOST,
}

/**
 * Satu unit berserial yang di-scan teknisi saat menyelesaikan work order.
 *
 * [serialNumber] dan [macAddress] SENGAJA disalin dari asetnya, bukan di-join belakangan:
 * ini catatan historis "apa yang terpasang di rumah pelanggan hari itu", dan tidak boleh
 * ikut berubah kalau baris asetnya kelak diperbaiki petugas master data.
 */
data class WorkOrderMaterialSerial(
    val id: UUID,
    val tenantId: UUID,
    val materialId: UUID,
    val workOrderId: UUID,
    val assetId: UUID,
    val serialNumber: String,
    val macAddress: String?,
    val outcome: MaterialOutcome,
    val scannedAt: Instant,
    val scannedBy: UUID,
) {
    init {
        require(serialNumber.isNotBlank()) { "nomor seri wajib diisi" }
    }
}

/**
 * Satu baris BOM: berapa banyak item X biasanya dipakai untuk jenis WO Y.
 *
 * BUKAN pembatas. Lihat komentar panjang di V186 — template hanya MEMPRA-ISI rencana, dan
 * yang dijaga sistem adalah selisihnya terlihat, bukan selisihnya dilarang.
 */
data class WorkOrderMaterialTemplateLine(
    val id: UUID,
    val tenantId: UUID,
    val workOrderType: String,
    val itemId: UUID,
    val plannedQuantity: Int,
    val note: String? = null,
) {
    init {
        require(workOrderType.isNotBlank()) { "jenis work order wajib diisi" }
        require(plannedQuantity > 0) { "jumlah template harus lebih dari nol" }
    }
}

/**
 * Rencana dan realisasi satu item pada satu work order.
 *
 * Invariant kuantitasnya adalah CERMIN dari CHECK `work_order_material_realized_ck` di V186.
 * Diduplikasi di domain SENGAJA: kegagalannya lebih berguna sebagai pesan Indonesia yang
 * bisa dibaca dispatcher daripada sebagai `ConstraintViolationException` mentah yang baru
 * meledak jauh di dalam flush Hibernate — di sana konteks "baris mana, item apa" sudah hilang.
 */
data class WorkOrderMaterialLine(
    val id: UUID,
    val tenantId: UUID,
    val workOrderId: UUID,
    val workOrderType: String,
    val itemId: UUID,
    val customerId: UUID,
    val itemCategory: String,
    val serialized: Boolean,
    val templateQuantity: Int?,
    val plannedQuantity: Int,
    val issuedQuantity: Int = 0,
    val usedQuantity: Int = 0,
    val returnedQuantity: Int = 0,
    val lostQuantity: Int = 0,
    val technicianId: UUID? = null,
    val technicianLocationId: UUID? = null,
    val varianceReason: String? = null,
    val serials: List<WorkOrderMaterialSerial> = emptyList(),
) {
    init {
        require(plannedQuantity >= 0 && issuedQuantity >= 0) { "jumlah material tidak boleh negatif" }
        require(usedQuantity >= 0 && returnedQuantity >= 0 && lostQuantity >= 0) { "jumlah material tidak boleh negatif" }
        require(usedQuantity + returnedQuantity + lostQuantity <= issuedQuantity) {
            "realisasi material melebihi yang dikeluarkan gudang"
        }
        require(issuedQuantity == 0 || (technicianId != null && technicianLocationId != null)) {
            "material yang sudah keluar gudang wajib punya pemegang"
        }
    }

    /** Ubah rencana. Tidak menyentuh realisasi — barang yang sudah keluar tidak bisa "direncanakan ulang". */
    fun replan(quantity: Int, template: Int?): WorkOrderMaterialLine {
        if (quantity < 0) throw ValidationException("Jumlah rencana tidak boleh negatif")
        if (quantity < issuedQuantity) {
            throw ConflictException("Rencana tidak bisa dikecilkan di bawah $issuedQuantity unit yang sudah dikeluarkan gudang")
        }
        return copy(plannedQuantity = quantity, templateQuantity = template ?: templateQuantity)
    }

    /**
     * Catat bahwa gudang benar-benar mengeluarkan [quantity] unit ke teknisi.
     *
     * Dipanggil SETELAH mutasi ISSUE berhasil, bukan sebelumnya: kalau angka ini naik lebih
     * dulu lalu mutasinya ditolak karena stok kurang, WO akan mengaku memegang barang yang
     * masih di rak.
     */
    fun markIssued(quantity: Int, technician: UUID, technicianLocation: UUID): WorkOrderMaterialLine {
        if (quantity <= 0) throw ValidationException("Jumlah pengeluaran harus lebih dari nol")
        if (technicianId != null && technicianId != technician) {
            // Satu baris material = satu dimensi saldo. Dua teknisi pada baris yang sama
            // membuat pemotongan saat approval mendarat di saldo orang yang salah, dan yang
            // satunya lagi tersisa stok hantu yang tidak pernah bisa diretur.
            throw ConflictException("Material item ini sudah dikeluarkan ke teknisi lain; buat baris terpisah")
        }
        return copy(
            issuedQuantity = issuedQuantity + quantity,
            technicianId = technician,
            technicianLocationId = technicianLocation,
        )
    }

    /**
     * Catat pemakaian barang CURAH (kabel, konektor, aksesori).
     *
     * Barang berserial DITOLAK di sini: jumlahnya diturunkan dari nomor seri yang di-scan,
     * bukan diketik. Kalau angkanya boleh diketik, "3 ONT terpasang" bisa berdiri tanpa satu
     * pun SN — dan saat pelanggan komplain, tidak ada yang bisa menjawab unit mana yang ada
     * di rumahnya.
     */
    fun recordBulkUsage(used: Int, returned: Int, lost: Int, reason: String?): WorkOrderMaterialLine {
        if (serialized) throw ValidationException("Item berserial: pemakaian dicatat lewat scan nomor seri, bukan diketik")
        if (used < 0 || returned < 0 || lost < 0) throw ValidationException("Jumlah pemakaian tidak boleh negatif")
        if (used + returned + lost > issuedQuantity) {
            throw ConflictException("Total realisasi ${used + returned + lost} melebihi $issuedQuantity unit yang dikeluarkan gudang")
        }
        return copy(
            usedQuantity = used,
            returnedQuantity = returned,
            lostQuantity = lost,
            varianceReason = reason?.trim()?.takeIf { it.isNotEmpty() } ?: varianceReason,
        )
    }

    /** Alasan selisih boleh diisi untuk baris apa pun, termasuk yang berserial. */
    fun explainVariance(reason: String): WorkOrderMaterialLine {
        if (reason.isBlank()) throw ValidationException("Alasan selisih tidak boleh kosong")
        return copy(varianceReason = reason.trim())
    }

    /**
     * Tempelkan hasil scan satu unit berserial dan turunkan ulang kuantitas realisasinya.
     *
     * Kuantitas SELALU dihitung ulang dari daftar serial, tidak pernah ditambah-kurangi
     * secara inkremental: penambahan inkremental akan melenceng diam-diam begitu satu scan
     * dikoreksi (mis. INSTALLED diubah jadi LOST), dan angka yang melenceng itulah yang
     * nanti dipakai memotong saldo.
     */
    fun attachSerial(serial: WorkOrderMaterialSerial): WorkOrderMaterialLine {
        if (!serialized) throw ValidationException("Item curah tidak punya nomor seri")
        if (serial.assetId !in serials.map { it.assetId } && serials.size >= issuedQuantity) {
            throw ConflictException("Sudah $issuedQuantity unit ter-scan, sebanyak yang dikeluarkan gudang")
        }
        val merged = serials.filterNot { it.assetId == serial.assetId } + serial
        return copy(
            serials = merged,
            usedQuantity = merged.count { it.outcome == MaterialOutcome.INSTALLED },
            returnedQuantity = merged.count { it.outcome == MaterialOutcome.RETURNED },
            lostQuantity = merged.count { it.outcome == MaterialOutcome.LOST },
        )
    }

    /** Unit berserial yang sudah keluar gudang tapi belum dilaporkan nasibnya. */
    val unscannedQuantity: Int get() = if (serialized) issuedQuantity - serials.size else 0

    /**
     * Penjaga penyelesaian WO. Melempar kalau baris ini belum layak ditutup.
     *
     * [itemLabel] hanya untuk pesan — teknisi di lapangan butuh tahu ITEM MANA yang kurang,
     * bukan uuid barisnya.
     */
    fun assertReadyForCompletion(itemLabel: String) {
        if (serialized) {
            if (unscannedQuantity > 0) {
                throw ValidationException(
                    "$itemLabel: $unscannedQuantity dari $issuedQuantity unit belum di-scan nomor serinya",
                )
            }
        } else if (usedQuantity + returnedQuantity + lostQuantity != issuedQuantity) {
            val sisa = issuedQuantity - (usedQuantity + returnedQuantity + lostQuantity)
            throw ValidationException("$itemLabel: $sisa unit belum dilaporkan terpakai, dikembalikan, atau hilang")
        }
        if (usedQuantity != plannedQuantity && varianceReason.isNullOrBlank()) {
            throw ValidationException(
                "$itemLabel: pemakaian $usedQuantity berbeda dari rencana $plannedQuantity — alasan selisih wajib diisi",
            )
        }
    }

    companion object {
        fun plan(
            tenantId: UUID,
            workOrderId: UUID,
            workOrderType: String,
            item: InventoryItem,
            customerId: UUID,
            quantity: Int,
            templateQuantity: Int?,
        ): WorkOrderMaterialLine {
            if (quantity < 0) throw ValidationException("Jumlah rencana tidak boleh negatif")
            return WorkOrderMaterialLine(
                id = UuidV7.generate(),
                tenantId = tenantId,
                workOrderId = workOrderId,
                workOrderType = workOrderType,
                itemId = item.id,
                customerId = customerId,
                itemCategory = item.category.name,
                serialized = item.serialized,
                templateQuantity = templateQuantity,
                plannedQuantity = quantity,
            )
        }
    }
}

File: web/src/api/workorderMaterial.ts
/**
 * Material satu work order: rencana (BOM), pengeluaran dari gudang, pemakaian di lapangan,
 * scan nomor seri, dan penarikan aset saat WO DISMANTLE.
 *
 * Terpisah dari `workorder.ts` karena pemiliknya memang modul lain: bentuk view-nya lahir di
 * `com.duluin.ftth.inventory.InventoryAllocationApi` dan `workorder` hanya meneruskannya.
 * Menyatukan keduanya membuat orang mengira `WorkOrderMaterialView` boleh diubah dari sisi
 * work order, padahal ia kontrak lintas modul.
 *
 * Barisnya SUDAH bernama lengkap (`itemName`, `technicianName`, `scannedByName`, …). JANGAN
 * gabungkan ulang `itemId`/`technicianId` ke `/item-master` atau `/api/users` di klien: dua
 * endpoint itu menuntut `inventory.item.view`/`iam.user.view` yang justru TIDAK dipegang
 * teknisi lapangan maupun petugas gudang biasa, dan penggabungan itulah yang dulu membuat
 * layar gudang memajang UUID telanjang.
 */

import { api } from './client'

/** GOOD = masih layak dipakai ulang, DAMAGED = rusak, masuk karantina saat WO disetujui. */
export type RecoveredAssetCondition = 'GOOD' | 'DAMAGED'

/**
 * Nasib satu unit berserial yang dibawa teknisi — WAJIB salah satu, tidak ada "tidak tahu".
 *
 * `INSTALLED` terpasang di pelanggan (satu-satunya yang memotong saldo saat WO disetujui),
 * `RETURNED` dibawa pulang utuh, `LOST` hilang ATAU rusak di lapangan. Sengaja TIDAK ada
 * `DAMAGED` terpisah: server hanya mengenal tiga, dan menawarkan pilihan keempat di layar
 * berarti teknisi memilih sesuatu yang dijawab 400 saat ia sudah berdiri di depan pelanggan.
 */
export type MaterialSerialOutcome = 'INSTALLED' | 'RETURNED' | 'LOST'

export interface WorkOrderMaterialSerialView {
  readonly id: string
  readonly assetId: string
  readonly serialNumber: string
  readonly macAddress: string | null
  readonly outcome: MaterialSerialOutcome
  readonly scannedAt: string
  readonly scannedBy: string
  readonly scannedByName: string
}

export interface WorkOrderMaterialView {
  readonly id: string
  readonly workOrderId: string
  readonly itemId: string
  readonly itemCode: string
  readonly itemName: string
  readonly itemCategory: string
  readonly unit: string
  readonly serialized: boolean
  /** Jumlah menurut BOM saat rencana dibuat; `null` berarti item di luar template. */
  readonly templateQuantity: number | null
  readonly plannedQuantity: number
  readonly issuedQuantity: number
  readonly usedQuantity: number
  readonly returnedQuantity: number
  readonly lostQuantity: number
  /** Unit berserial yang sudah keluar gudang tapi belum di-scan nasibnya. */
  readonly unscannedQuantity: number
  readonly technicianId: string | null
  readonly technicianName: string | null
  readonly technicianLocationId: string | null
  readonly varianceReason: string | null
  readonly serials: readonly WorkOrderMaterialSerialView[]
}

export interface WorkOrderMaterialTemplateView {
  readonly itemId: string
  readonly itemCode: string
  readonly itemName: string
  readonly itemCategory: string
  readonly unit: string
  readonly serialized: boolean
  readonly plannedQuantity: number
  readonly note: string | null
}

export interface WorkOrderRecoveredAssetView {
  readonly id: string
  readonly workOrderId: string
  readonly assetId: string
  readonly serialNumber: string
  readonly macAddress: string | null
  readonly itemId: string
  readonly itemCode: string
  readonly itemName: string
  readonly itemCategory: string
  readonly customerId: string
  readonly technicianId: string
  readonly technicianName: string
  readonly technicianLocationId: string
  readonly condition: RecoveredAssetCondition
  readonly note: string | null
  readonly recoveredAt: string
  readonly recoveredBy: string
  readonly recoveredByName: string
  /** Terisi berarti barisnya DIBATALKAN — barisnya tetap ada, sengaja tidak dihapus. */
  readonly cancelledAt: string | null
  readonly cancelledBy: string | null
  readonly cancelledByName: string | null
  readonly cancelReason: string | null
}

export interface PlanMaterialLineBody {
  readonly itemId: string
  readonly quantity: number
}

export interface IssueMaterialLineBody {
  readonly itemId: string
  readonly quantity: number
  readonly serialNumbers?: readonly string[]
}

export interface IssueMaterialBody {
  readonly fromLocationId: string
  readonly custodianId: string
  readonly technicianId: string
  readonly technicianLocationId: string
  readonly lines: readonly IssueMaterialLineBody[]
  readonly reason: string
  readonly operationKey: string
  readonly payloadHash: string
}

export interface MaterialUsageBody {
  readonly itemId: string
  readonly usedQuantity: number
  readonly returnedQuantity?: number
  readonly lostQuantity?: number
  readonly varianceReason?: string | null
}

export interface MaterialSerialBody {
  readonly serialNumber: string
  readonly outcome: MaterialSerialOutcome
  readonly macAddress?: string | null
}

export interface RecoverAssetBody {
  readonly serialNumber: string
  readonly technicianId: string
  readonly technicianLocationId: string
  readonly condition: RecoveredAssetCondition
  readonly note?: string | null
}

const base = (workOrderId: string) => `/api/work-orders/${workOrderId}/materials`

export const listWorkOrderMaterials = (workOrderId: string) =>
  api.get<WorkOrderMaterialView[]>(base(workOrderId))

export const getWorkOrderMaterialTemplate = (workOrderId: string) =>
  api.get<WorkOrderMaterialTemplateView[]>(`${base(workOrderId)}/template`)

/**
 * Simpan rencana material. `lines: []` berarti "pakai BOM apa adanya" — itulah jalur pra-isi
 * yang dipakai dispatcher saat membuka WO baru, bukan cara mengosongkan rencana.
 */
export const planWorkOrderMaterial = (workOrderId: string, lines: readonly PlanMaterialLineBody[]) =>
  api.put<WorkOrderMaterialView[]>(base(workOrderId), { lines })

export const issueWorkOrderMaterial = (workOrderId: string, body: IssueMaterialBody) =>
  api.post<WorkOrderMaterialView[]>(`${base(workOrderId)}/issues`, body)

/** Pemakaian barang CURAH. Barang berserial ditolak server — jumlahnya datang dari scan. */
export const recordWorkOrderMaterialUsage = (workOrderId: string, body: MaterialUsageBody) =>
  api.post<WorkOrderMaterialView>(`${base(workOrderId)}/usage`, body)

export const scanWorkOrderMaterialSerial = (workOrderId: string, body: MaterialSerialBody) =>
  api.post<WorkOrderMaterialView>(`${base(workOrderId)}/serials`, body)

export const listRecoveredAssets = (workOrderId: string) =>
  api.get<WorkOrderRecoveredAssetView[]>(`${base(workOrderId)}/recovered-assets`)

export const recoverWorkOrderAsset = (workOrderId: string, body: RecoverAssetBody) =>
  api.post<WorkOrderRecoveredAssetView>(`${base(workOrderId)}/recovered-assets`, body)

/**
 * POST, bukan DELETE: barisnya TIDAK dihapus. Jejak "pernah tercatat ditarik lalu dibatalkan,
 * oleh siapa, dengan alasan apa" justru bagian yang paling perlu dibaca saat menyelisik selisih
 * stok. DELETE akan membuat pembacanya mengira barisnya lenyap.
 */
export const cancelRecoveredAsset = (workOrderId: string, recoveredId: string, reason: string | null) =>
  api.post<WorkOrderRecoveredAssetView>(`${base(workOrderId)}/recovered-assets/${recoveredId}/cancel`, { reason })
